File: onion-architecture/java/src/test/java/com/example/domain/CourseTest.java
package com.example.domain;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.Test;

class CourseTest {

    @Test
    void shouldCreateCourseWithIdAndTitle() {
        Long id = 1L;
        String title = "Java Programming";

        Course course = new Course(id, title);

        assertEquals(id, course.getId());
        assertEquals(title, course.getTitle());
    }

    @Test
    void shouldReturnTrueForEqualCoursesWithSameId() {
        Course course1 = new Course(1L, "Java Programming");
        Course course2 = new Course(1L, "Different Title");

        assertTrue(course1.equals(course2));
        assertEquals(course1.hashCode(), course2.hashCode());
    }

    @Test
    void shouldReturnFalseForCoursesWithDifferentIds() {
        Course course1 = new Course(1L, "Java Programming");
        Course course2 = new Course(2L, "Java Programming");

        assertFalse(course1.equals(course2));
        assertNotEquals(course1.hashCode(), course2.hashCode());
    }

    @Test
    void shouldReturnFalseWhenComparingWithNull() {
        Course course = new Course(1L, "Java Programming");

        assertFalse(course.equals(null));
    }

    @Test
    void shouldReturnFalseWhenComparingWithDifferentType() {
        Course course = new Course(1L, "Java Programming");
        String notACourse = "Not a course";

        assertFalse(course.equals(notACourse));
    }

    @Test
    void shouldReturnTrueWhenComparingWithItself() {
        Course course = new Course(1L, "Java Programming");

        assertTrue(course.equals(course));
    }

    @Test
    void shouldHaveConsistentHashCode() {
        Course course = new Course(1L, "Java Programming");
        int initialHashCode = course.hashCode();

        assertEquals(initialHashCode, course.hashCode());
        assertEquals(initialHashCode, course.hashCode());
    }
}

File: onion-architecture/kotlin/src/test/kotlin/com/example/domain/CourseTest.kt
package com.example.domain

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*

class CourseTest {

    @Test
    fun shouldCreateCourseWithIdAndTitle() {
        val id = 1L
        val title = "Java Programming"
        
        val course = Course(id, title)
        
        assertEquals(id, course.id)
        assertEquals(title, course.title)
    }

    @Test
    fun shouldReturnTrueForEqualCoursesWithSameId() {
        val course1 = Course(1L, "Java Programming")
        val course2 = Course(1L, "Different Title")
        
        assertTrue(course1 == course2)
        assertEquals(course1.hashCode(), course2.hashCode())
    }

    @Test
    fun shouldReturnFalseForCoursesWithDifferentIds() {
        val course1 = Course(1L, "Java Programming")
        val course2 = Course(2L, "Java Programming")
        
        assertFalse(course1 == course2)
        assertNotEquals(course1.hashCode(), course2.hashCode())
    }

    @Test
    fun shouldReturnFalseWhenComparingWithNull() {
        val course = Course(1L, "Java Programming")
        
        assertFalse(course == null)
    }

    @Test
    fun shouldReturnFalseWhenComparingWithDifferentType() {
        val course = Course(1L, "Java Programming")
        val notACourse = "Not a course"
        
        assertFalse(course.equals(notACourse))
    }

    @Test
    fun shouldReturnTrueWhenComparingWithItself() {
        val course = Course(1L, "Java Programming")
        
        assertTrue(course == course)
    }

    @Test
    fun shouldHaveConsistentHashCode() {
        val course = Course(1L, "Java Programming")
        val initialHashCode = course.hashCode()
        
        assertEquals(initialHashCode, course.hashCode())
        assertEquals(initialHashCode, course.hashCode())
    }
}